### PR TITLE
policy: add CLI for introspection into state of a SelectorCache

### PR DESCRIPTION
**Pull request by ianvernon**
This PR adds the required code in various parts of Cilium (API models, Daemon handling of API request, and creation of API model from SelectorCache) to allow end-users to introspect the state of the SelectorCache which will be used for policy computation:

Example output:

```
$ cilium policy cache 
SELECTOR                                                                                                                                  USERS   IDENTITIES
&LabelSelector{MatchLabels:map[string]string{k8s.io.kubernetes.pod.namespace: kube-system,k8s.k8s-app: kube-dns,},MatchExpressions:[],}   2       102    
                                                                                                                                                  103    
                                                                                                                                                  104    
                                                                                                                                                  106    
&LabelSelector{MatchLabels:map[string]string{k8s.io.kubernetes.pod.namespace: default,k8s.org: empire,},MatchExpressions:[],}             3       8033 
```

* SELECTOR represents the string form of the selector in the cache.
* USERS represents the number of 'users' of this selector.
* IDENTITIES is the set of security identities in their numeric representation which match this selector.

Signed-off by: Ian Vernon <ian@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8024)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

File: Documentation/cmdref/cilium_policy.md [hunk 1]
```diff
@@ -25,6 +25,7 @@ Manage security policies
 ### SEE ALSO
 
 * [cilium](../cilium)	 - CLI
+* [cilium policy cache](../cilium_policy_cache)	 - Display cache mapping selectors to identities
```
Comment (jrajahalme) [inline Documentation/cmdref/cilium_policy.md]: We also have a PolicyCache, should we name this API something different? E.g., reserve "cilium policy cache" for the PolicyCache introspection, and maybe "cilium policy selectors" for this one?

File: pkg/policy/selectorcache.go [hunk 2]
```diff
@@ -129,6 +130,7 @@ type identitySelector interface {
 	addUser(CachedSelectionUser) (added bool)
 	removeUser(CachedSelectionUser) (last bool)
 	notifyUsers(added, deleted []identity.NumericIdentity)
+	numUsers() int
```
Comment (jrajahalme) [inline pkg/policy/selectorcache.go:133]: IMO this is OK for now, but ee could also add a call to the Users interface to return a friendly string name for the user, then add a call here to return a slice of those names? Then we could show the list of users (selector policies).

**Review by jrajahalme — APPROVED** [review]
Looking good, small nit only. And maybe document that this API is subject to change between Cilium versions (at least for the time being)?

File: pkg/policy/selectorcache.go [hunk 3]
```diff
@@ -172,6 +174,30 @@ type SelectorCache struct {
 	selectors map[string]identitySelector
 }
 
+// GetModel returns the API model of the SelectorCache.
+func (sc *SelectorCache) GetModel() models.SelectorCache {
+	sc.mutex.Lock()
```
Comment (tgraf) [inline pkg/policy/selectorcache.go]: RLock?

**Review by tgraf — APPROVED** [review]
Love it, very nice! We might need slightly more intelligent printing around the list of identities as we encounter systems with thousands of identities.

**Review by aanm — COMMENTED** [review]
Is the example output correct? I can't see where the "`&`" comes from?

**Comment by ianvernon** [comment]
@aanm we get the string representation from the `String()` function of the `LabelSelector` within the `EndpointSelector`:

https://github.com/kubernetes/apimachinery/blob/b74e4c97951f106696db86d7593effcb1463ca30/pkg/apis/meta/v1/generated.pb.go#L2895-L2915

**Comment by ianvernon** [comment]
test-me-please

File: Documentation/cmdref/cilium_policy.md (hunk 1 above)
Comment (ianvernon) [reply Documentation/cmdref/cilium_policy.md]: Done.

File: pkg/policy/selectorcache.go (hunk 2 above)
Comment (ianvernon) [reply pkg/policy/selectorcache.go:133]: Sure, we can add these as time goes on. I think right now this is enough for debugging refcount issues and seeing whether stale identities are left behind, identities are mapped correctly, etc.

**Comment by ianvernon** [comment]
test-me-please

**Comment by jrfastab** [comment]
LGTM

**Pull request merged**